File: pbweb/pbweb-jar/src/main/java/it/csi/pbandi/pbweb/business/service/FornitoreService.java
package it.csi.pbandi.pbweb.business.service;

import java.security.InvalidParameterException;

import javax.servlet.http.HttpServletRequest;
import javax.ws.rs.Consumes;
import javax.ws.rs.DefaultValue;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.QueryParam;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import it.csi.pbandi.pbweb.dto.FornitoreFormDTO;
//import it.csi.pbandi.pbweb.integration.dao.request.AssociaFilesRequest;
import it.csi.pbandi.pbservizit.integration.dao.request.AssociaFilesRequest;
import it.csi.pbandi.pbweb.integration.dao.request.SalvaQualificaRequest;
import it.csi.pbandi.pbweb.integration.dao.request.SalvaQualificheRequest;
import it.csi.pbandi.pbweb.pbandisrv.dto.gestionefornitori.FornitoreDTO;

@Path("/fornitore")
public interface FornitoreService {
	
	@GET
	@Path("testTransactional") 
	@Produces(MediaType.APPLICATION_JSON)
	Response testTransactional(@Context HttpServletRequest req) throws RuntimeException, InvalidParameterException, Exception;
	
	@GET
	@Path("alberoAttivitaAteco") 
	@Produces(MediaType.APPLICATION_JSON)
	Response alberoAttivitaAteco(
			@DefaultValue("0") @QueryParam("idUtente") long idUtente,
			@Context HttpServletRequest req) throws InvalidParameterException, Exception;
	
	@GET
	@Path("attivitaAteco") 
	@Produces(MediaType.APPLICATION_JSON)
	Response attivitaAteco(
			@DefaultValue("0") @QueryParam("idUtente") long idUtente,
			@Context HttpServletRequest req) throws InvalidParameterException, Exception;
	
	@POST
	@Path("salvaFornitore")
	@Consumes(MediaType.APPLICATION_JSON)
	@Produces(MediaType.APPLICATION_JSON)
	Response salvaFornitore(
			FornitoreFormDTO fornitoreFormDTO,
			@DefaultValue("0") @QueryParam("idUtente") long idUtente,
			@DefaultValue("0") @QueryParam("idSoggettoBeneficiario") long idSoggettoBeneficiario,
			@Context HttpServletRequest request) 
			throws InvalidParameterException, Exception;

	@GET
	@Path("cercaFornitore") 
	@Produces(MediaType.APPLICATION_JSON)
	Response cercaFornitore(
			@QueryParam("idFornitore") Long idFornitore,
			@QueryParam("idSoggettoBeneficiario") Long idSoggettoBeneficiario,
			@QueryParam("idProgetto") Long idProgetto,
			@Context HttpServletRequest req) throws InvalidParameterException, Exception;

	@GET
	@Path("qualificheFornitore")
	@Produces(MediaType.APPLICATION_JSON)
	Response qualificheFornitore(
			@DefaultValue("0") @QueryParam("idFornitore") long idFornitore,
			@DefaultValue("0") @QueryParam("idUtente") long idUtente,
			@Context HttpServletRequest req) throws InvalidParameterException, Exception;
	
	@POST
	@Path("salvaQualifica")
	@Consumes(MediaType.APPLICATION_JSON)
	@Produces(MediaType.APPLICATION_JSON)
	Response salvaQualifica(
			SalvaQualificaRequest salvaQualificaRequest,
			@Context HttpServletRequest request) 
			throws InvalidParameterException, Exception;
	
	@POST
	@Path("salvaQualifiche")
	@Consumes(MediaType.APPLICATION_JSON)
	@Produces(MediaType.APPLICATION_JSON)
	Response salvaQualifiche(
			SalvaQualificheRequest salvaQualificheRequest,
			@Context HttpServletRequest request) 
			throws InvalidParameterException, Exception;
	
	@GET
	@Path("fattureFornitore")
	@Produces(MediaType.APPLICATION_JSON)
	Response fattureFornitore(
			@DefaultValue("0") @QueryParam("idFornitore") long idFornitore,
			@DefaultValue("0") @QueryParam("idProgetto") long idProgetto,
			@Context HttpServletRequest req) throws InvalidParameterException, Exception;
	
	@GET
	@Path("nazioni")
	@Produces(MediaType.APPLICATION_JSON)
	Response nazioni(
			@Context HttpServletRequest req) throws InvalidParameterException, Exception;
	
	@POST
	@Path("ricercaElencoFornitori")
	@Consumes(MediaType.APPLICATION_JSON)
	@Produces(MediaType.APPLICATION_JSON)
	Response ricercaElencoFornitori(
			FornitoreDTO filtro,
			@QueryParam("idSoggettoBeneficiario") Long idSoggettoBeneficiario,
			@QueryParam("idProgetto") Long idProgetto,
			@Context HttpServletRequest request) throws InvalidParameterException, Exception;
	
	@GET
	@Path("disassociaAllegatoFornitore")
	@Produces(MediaType.APPLICATION_JSON)
	Response disassociaAllegatoFornitore(
			@QueryParam("idDocumentoIndex") Long idDocumentoIndex, 
			@QueryParam("idFornitore") Long idFornitore, 
			@QueryParam("idProgetto") Long idProgetto,
			@Context HttpServletRequest req) throws InvalidParameterException, Exception;
	
	@POST
	@Path("associaAllegatiAFornitore")
	@Consumes(MediaType.APPLICATION_JSON)
	@Produces(MediaType.APPLICATION_JSON)
	Response associaAllegatiAFornitore(
			AssociaFilesRequest associaFilesRequest,
			@Context HttpServletRequest request) throws InvalidParameterException, Exception;
	
	@GET
	@Path("eliminaFornitore")
	@Produces(MediaType.APPLICATION_JSON)
	Response eliminaFornitore(
			@QueryParam("idFornitore") Long idFornitore, 
			@Context HttpServletRequest req) throws InvalidParameterException, Exception;
	
	@GET
	@Path("esistonoDocumentiAssociatiAQualificaFornitore")
	@Produces(MediaType.APPLICATION_JSON)
	Response esistonoDocumentiAssociatiAQualificaFornitore(
			@QueryParam("progrFornitoreQualifica") Long progrFornitoreQualifica, 
			@Context HttpServletRequest req) throws InvalidParameterException, Exception;
	
	@GET
	@Path("eliminaQualifica")
	@Produces(MediaType.APPLICATION_JSON)
	Response eliminaQualifica(
			@QueryParam("progrFornitoreQualifica") Long progrFornitoreQualifica, 
			@Context HttpServletRequest req) throws InvalidParameterException, Exception;
}

File: pbwebwcl/src/app/rendicontazione/services/fornitore.service.ts
/*
 * Copyright Regione Piemonte - 2024 
 * SPDX-License-Identifier: EUPL-1.2 
*/

import { Observable } from 'rxjs';
import { SalvaQualificaRequest } from '../commons/requests/salva-qualifica-request';
import { UserInfoSec } from '../../core/commons/dto/user-info';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { ConfigService } from 'src/app/core/services/config.service';
import { QualificaDTO } from '../commons/dto/qualifica-dto';
import { FornitoreFormDTO } from '../commons/dto/fornitore-form-dto';
import { EsitoDTO } from 'src/app/shared/commons/dto/esito-dto';
import { AttivitaAtecoNodoDTO } from '../commons/dto/attivita-ateco-nodo-dto';
import { NazioneDTO } from '../commons/dto/nazione-dto';
import { FatturaRiferimentoDTO } from '../commons/dto/fattura-riferimento.dto';
import { FornitoreDTO } from '../commons/dto/fornitore-dto';
import { EsitoRicercaFornitori } from '../commons/dto/esito-ricerca-fornitori';
import { SalvaQualificheRequest } from '../commons/requests/salva-qualifiche-request';
import { AssociaFilesRequest } from '../commons/requests/associa-files-request';
import { EsitoAssociaFilesDTO } from '../commons/dto/esito-associa-files-dto';
import { EsitoOperazioneDTO } from 'src/app/shared/commons/dto/esito-operazione-dto';

@Injectable()
export class FornitoreService {

    constructor(
        private configService: ConfigService,
        private http: HttpClient
    ) { }

    cercaFornitore(idFornitore: number, idProgetto: number, idSoggettoBeneficiario: number): Observable<FornitoreFormDTO> {
        let url = this.configService.getApiURL() + '/restfacade/fornitore/cercaFornitore';
        let params = new HttpParams()
            .set("idFornitore", idFornitore.toString())
            .set("idProgetto", idProgetto.toString())
            .set("idSoggettoBeneficiario", idSoggettoBeneficiario.toString());
        return this.http.get<FornitoreFormDTO>(url, { params: params });
    }

    salvaFornitore(fornitoreFormDTO: FornitoreFormDTO, idUtente: number, idSoggettoBeneficiario: number): Observable<EsitoDTO> {
        let url = this.configService.getApiURL() + '/restfacade/fornitore/salvaFornitore';
        let params = new HttpParams().set("idUtente", idUtente.toString()).set("idSoggettoBeneficiario", idSoggettoBeneficiario.toString());
        return this.http.post<EsitoDTO>(url, fornitoreFormDTO, { params: params });
    }

    salvaQualifica(request: SalvaQualificaRequest): Observable<EsitoDTO> {
        let url = this.configService.getApiURL() + '/restfacade/fornitore/salvaQualifica';
        return this.http.post<EsitoDTO>(url, request);
    }

    getQualificheFornitore(userInfo: UserInfoSec, idFornitore: number) {
        let url = this.configService.getApiURL() + '/restfacade/fornitore/qualificheFornitore';

        let params = new HttpParams()
            .set("idFornitore", idFornitore.toString())
            .set("idUtente", userInfo.idUtente.toString())

        return this.http.get<Array<QualificaDTO>>(url, { params: params });
    }

    getFattureFornitore(idProgetto: number, idFornitore: number): Observable<Array<FatturaRiferimentoDTO>> {
        let url = this.configService.getApiURL() + '/restfacade/fornitore/fattureFornitore';
        let params = new HttpParams().set("idFornitore", idFornitore.toString()).set("idProgetto", idProgetto.toString());
        return this.http.get<Array<FatturaRiferimentoDTO>>(url, { params: params });
    }

    getAlberoAttivitaAteco(idUtente: number): Observable<Array<AttivitaAtecoNodoDTO>> {
        let url = this.configService.getApiURL() + '/restfacade/fornitore/alberoAttivitaAteco';
        let params = new HttpParams().set("idUtente", idUtente.toString());
        return this.http.get<Array<AttivitaAtecoNodoDTO>>(url, { params: params });
    }

    getNazioni(): Observable<Array<NazioneDTO>> {
        let url = this.configService.getApiURL() + '/restfacade/fornitore/nazioni';
        return this.http.get<Array<NazioneDTO>>(url);
    }

    ricercaElencoFornitori(filtro: FornitoreDTO, idSoggettoBeneficiario: number, idProgetto: number): Observable<EsitoRicercaFornitori> {
        let url = this.configService.getApiURL() + '/restfacade/fornitore/ricercaElencoFornitori';
        let params = new HttpParams().set("idSoggettoBeneficiario", idSoggettoBeneficiario.toString()).set("idProgetto", idProgetto.toString());
        return this.http.post<EsitoRicercaFornitori>(url, filtro, { params: params });
    }

    salvaQualifiche(request: SalvaQualificheRequest): Observable<EsitoDTO> {
        let url = this.configService.getApiURL() + '/restfacade/fornitore/salvaQualifiche';
        return this.http.post<EsitoDTO>(url, request);
    }

    associaAllegatiAFornitore(request: AssociaFilesRequest): Observable<EsitoAssociaFilesDTO> {
        let url = this.configService.getApiURL() + '/restfacade/fornitore/associaAllegatiAFornitore';
        return this.http.post<EsitoAssociaFilesDTO>(url, request);
    }

    disassociaAllegatoFornitore(idDocumentoIndex: number, idFornitore: number, idProgetto: number): Observable<EsitoDTO> {
        let url = this.configService.getApiURL() + '/restfacade/fornitore/disassociaAllegatoFornitore';
        let params = new HttpParams().set("idDocumentoIndex", idDocumentoIndex.toString()).set("idFornitore", idFornitore.toString()).set("idProgetto", idProgetto.toString());
        return this.http.get<EsitoDTO>(url, { params: params });
    }

    esistonoDocumentiAssociatiAQualificaFornitore(progrFornitoreQualifica: number): Observable<boolean> {
        let url = this.configService.getApiURL() + '/restfacade/fornitore/esistonoDocumentiAssociatiAQualificaFornitore';
        let params = new HttpParams().set("progrFornitoreQualifica", progrFornitoreQualifica.toString());
        return this.http.get<boolean>(url, { params: params });
    }

    eliminaQualifica(progrFornitoreQualifica: number): Observable<EsitoDTO> {
        let url = this.configService.getApiURL() + '/restfacade/fornitore/eliminaQualifica';
        let params = new HttpParams().set("progrFornitoreQualifica", progrFornitoreQualifica.toString());
        return this.http.get<EsitoDTO>(url, { params: params });
    }

    eliminaFornitore(idFornitore: number): Observable<EsitoOperazioneDTO> {
        let url = this.configService.getApiURL() + '/restfacade/fornitore/eliminaFornitore';
        let params = new HttpParams().set("idFornitore", idFornitore.toString());
        return this.http.get<EsitoOperazioneDTO>(url, { params: params });
    }
}
